network/rcmgr.go

```diff
@@ -0,0 +1,116 @@
+package network
+
+import (
+	"errors"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var (
+	ErrResourceLimitExceeded = errors.New("resource limit exceeded")
+	ErrResourceScopeClosed   = errors.New("resource scope closed")
+)
+
+// ResourceManager is the interface to the network resource management subsystem
+type ResourceManager interface {
+	// GetProtocolScope retrieves the resource management scope for a specific protocol.
+	// If there is no configured limits for a particular protocol, then the default scope is
+	// returned.
+	GetProtocolScope(protocol.ID) ProtocolScope
+	// GetPeerScope retrieces the resource management scope for a specific peer.
+	GetPeerScope(peer.ID) PeerScope
+
+	// OpenConnection creates a connection scope not yet associated with any peer
+	OpenConnection(dir Direction, usefd bool) (ConnectionScope, error)
+
+	// Close closes the resource manager
+	Close() error
+}
+
+// ResourceScope is the interface for all scopes.
+type ResourceScope interface {
+	// ReserveMemory reserves memory/buffer space in the scope.
+	ReserveMemory(size int) error
+	// ReleaseMemory explicitly releases memory previously reserved with ReserveMemory
+	ReleaseMemory(size int)
+
+	// GetBuffer reserves memory and allocates a buffer through the buffer pool.
+	GetBuffer(size int) ([]byte, error)
+	// GrowBuffer grows a previous allocated buffer, reserving the appropriate memory space.
+	GrowBuffer(buf []byte, newsize, copy int) ([]byte, error)
+	// ReleaseBuffer releases a previous allocated buffer.
+	ReleaseBuffer(buf []byte)
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
+}
+
+// TransactionalScope is a mixin interface for transactional scopes.
+type TransactionalScope interface {
+	// Done ends the transaction scope and releases associated resources.
+	Done()
+}
+
+// ProtocolScope is the interface for protocol resource scopes.
+type ProtocolScope interface {
+	ResourceScope
+
+	// Protocols returns the list of protocol IDs constrained by this scope.
+	Protocols() []protocol.ID
+	// IsDefault returns true if this is the default (global) scope.
+	IsDefault() bool
```

Comment:
What would be a non-default scope?

Reply:
I removed this method, it doesn't make much sense.